xclim/indices/_agro.py

```diff
@@ -884,28 +885,97 @@ def water_budget(
     return out
 
 
+def _preprocess_spx(da, freq, window, **indexer):
+    _, base, _, _ = parse_offset(freq or xarray.infer_freq(da.time))
+    try:
+        group = {"D": "time.dayofyear", "M": "time.month"}[base]
+    except KeyError():
+        raise ValueError(f"Standardized index with frequency `{freq}` not supported.")
+    if freq:
+        da = da.resample(time=freq).mean(keep_attrs=True)
+    if window > 1:
+        da = da.rolling(time=window).mean(skipna=False, keep_attrs=True)
+    da = select_time(da, **indexer)
+    return da, group
+
+
+def _compute_spx_fit_params(
+    da, cal_range, freq, window, dist, method, group=None, **indexer
+):
+    # "WPM" method doesn't seem to work for gamma or pearson3
+    dist_and_methods = {"gamma": ["ML", "APP"], "fisk": ["ML", "APP"]}
+    if dist not in dist_and_methods:
+        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
+    if method not in dist_and_methods[dist]:
+        raise NotImplementedError(
+            f"The method `{method}` is not supported for distribution `{dist}`."
+        )
+
+    if group is None:
+        da, group = _preprocess_spx(da, freq, window)
+
+    if cal_range:
+        da = da.sel(time=slice(cal_range[0], cal_range[1]))
+
+    if uses_dask(da) and len(da.chunks[da.get_axis_num("time")]) > 1:
+        warnings.warn(
+            "The input data is chunked on time dimension and must be fully rechunked to"
+            " run `fit` on groups ."
+            " Beware, this operation can significantly increase the number of tasks dask"
+            " has to handle.",
+            stacklevel=2,
+        )
+        da = da.chunk({"time": -1})
+
+    da = select_time(da, **indexer)
+
+    def wrap_fit(da):
+        if indexer != {}:
+            if da.isnull.all():
+                select_dims = {d: 0 for d in da.dims if d != "time"}
+                with xarray.set_options(keep_attrs=True):
+                    params = (
+                        fit(da.isel(time=slice(0, 2))[select_dims], dist, method)
+                        * da.isel(time=0, drop=True)
+                        * np.NaN
+                    )
+                return params
```

Comment:
I thought that for periods of time that have been cut by indexing, we should not really compute fitting parameters (or spi values later) we just fill in with NaN. I still run it through the fitting pipeline so that I have appropriate dimensions. I broadcast the result to the full `da`. 

Perhaps this logic could be used more widely or better formulated. I wonder if a trick like this could speed-up the computation for all-NaN slices, or if scipy already makes such checks.